src/integrationTest/java/uk/gov/hmcts/reform/wacaseeventhandler/UpdateRecordErrorHandlingTest.java

```diff
@@ -0,0 +1,163 @@
+            "classpath:sql/insert_case_event_messages_for_processing_from_dlq.sql"})
+    @Test
+    void should_set_message_state_to_processed_when_message_update_failed_in_first_time()
+        throws JsonProcessingException {
+        doNothing().when(ccdEventProcessor).processMessage(any(CaseEventMessage.class));
+        doThrow(new TransactionTimedOutException("Time out")).when(caseEventMessageRepository)
+            .updateMessageState(eq(MessageState.PROCESSED), Mockito.<String>anyList());
+        await()
+            .atMost(60, SECONDS)
+            .untilAsserted(() -> {
+                        assertEquals(1, getMessagesInDbFromQuery(PROCESSED_STATE_QUERY).size());
+                    }
+            );
+    }
+
+    @Sql(executionPhase = Sql.ExecutionPhase.BEFORE_TEST_METHOD,
+        scripts = {"classpath:sql/delete_from_case_event_messages.sql",
+            "classpath:sql/insert_case_event_messages_for_processing_ready_msgs.sql"})
+    @Test
+    void should_update_hold_until_and_retry_count_for_ready_messages_when_retryable_exception_occurs()
+        throws JsonProcessingException {
+        String caseId = "6761065058131570";
+
+        MessageProcessorTest.RetryableFeignException retryableFeignException = new MessageProcessorTest
+            .RetryableFeignException(504, "Gateway Timeout");
+        doThrow(retryableFeignException)
+            .when(ccdEventProcessor)
+            .processMessage(any(CaseEventMessage.class));
+        doThrow(new TransactionTimedOutException("Time out")).when(caseEventMessageRepository)
+            .updateMessageWithRetryDetails(eq(1), any(LocalDateTime.class), eq(MESSAGE_ID));
+
+        await()
+            .atMost(60, SECONDS)
+            .untilAsserted(() -> {
+                    assertEquals(2, getMessagesInDbFromQuery(format("case_id=%s", caseId)).size());
+                    assertEquals(1, getMessageById(MESSAGE_ID).getRetryCount());
+                    assertNotNull(getMessageById(MESSAGE_ID).getHoldUntil());
+                    }
+            );
+    }
+
+    @Sql(executionPhase = Sql.ExecutionPhase.BEFORE_TEST_METHOD,
+        scripts = {"classpath:sql/delete_from_case_event_messages.sql",
+            "classpath:sql/insert_case_event_messages_for_processing_ready_msgs.sql"})
+    @Test
+    void should_set_message_state_to_unprocessable_when_non_retryable_error_occurs() throws JsonProcessingException {
+        doThrow(FeignException.NotFound.class).when(ccdEventProcessor).processMessage(any(CaseEventMessage.class));
+        doThrow(new TransactionTimedOutException("Time out")).when(caseEventMessageRepository)
+            .updateMessageState(eq(MessageState.UNPROCESSABLE), Mockito.<String>anyList());
+        await()
+            .atMost(60, SECONDS)
+            .untilAsserted(() -> {
+                    assertEquals(1, getMessagesInDbFromQuery(UNPROCESSABLE_STATE_QUERY).size());
+                    }
+            );
+    }
+
```

Comment:
do we have tests to cover;
1. When state is ready and hold_until in future, update to new state?
2. When state is ready and hold_until in past, update to new state?